anta/runner.py

```diff
@@ -77,6 +104,61 @@ def log_cache_statistics(devices: list[AntaDevice]) -> None:
             logger.info("Caching is not enabled on %s", device.name)
 
 
+async def run(tests_generator: AsyncGenerator[Coroutine[Any, Any, TestResult], None], limit: int) -> AsyncGenerator[TestResult, None]:
+    """Run tests with a concurrency limit.
+
+    This function takes an asynchronous generator of test coroutines and runs them
+    with a limit on the number of concurrent tests. It yields test results as each
+    test completes.
+
+    Inspired by: https://death.andgravity.com/limit-concurrency
+
+    Parameters
+    ----------
+    tests_generator
+        An asynchronous generator that yields test coroutines.
+    limit
+        The maximum number of concurrent tests to run.
+
+    Yields
+    ------
+        The result of each completed test.
+    """
+    # NOTE: The `aiter` built-in function is not available in Python 3.9
```

Comment:
so this will fail in 3.9?

Reply:
`aiter()` and `anext()` were released in 3.10: https://docs.python.org/3/library/functions.html#aiter so we need to use the dunders instead (since we support 3.9 RIP)

Reply:
not RIP yet: https://devguide.python.org/versions/

Reply:
ok so this can be like this until we drop 3.9 understood thanks!